Support native parquet stores with scalar-only Parquet types

selitvin opened the pull request:
Adding `make_parquet_reader`: support for working efficiently with parquet stores.
    
Currently only stores with scalars are supported.
Readers created with `make_parquet_reader` return batches of rows and not row-by-row, as `make_reader`.

rgruener commented on petastorm/tests/test_parquet_reader.py:
```diff
@@ -0,0 +1,57 @@
+    with reader_factory(scalar_dataset.url) as reader:
+        _check_simple_reader(reader, scalar_dataset.data)
+
+# TODO(yevgeni): missing tests: subset of columns; invalid column name;
```
Maybe attach a github issue to this todo?

rgruener commented on petastorm/tests/test_end_to_end_predicates_impl.py:
```diff
@@ -37,3 +37,21 @@ def get_fields(self):
+        return list(self._values.keys())
+
+    def do_include(self, values):
+        result = None
```
It might be more concise/clear to write this as:
```
        result = [True] * len(values)
        for field_name in self._values.keys():
            result &= values[field_name] == self._values[field_name]
        return result
```

rgruener replied:
Though actually that might just be my functional programming preference actually

rgruener commented on petastorm/tests/test_end_to_end.py:
```diff
@@ -89,24 +119,24 @@ def test_simple_read_with_disk_cache(synthetic_dataset, reader_factory, tmpdir):
                         cache_type='local-disk', cache_location=tmpdir.strpath,
                         cache_size_limit=CACHE_SIZE, cache_row_size_estimate=ROW_SIZE_BYTES) as reader:
-        _check_simple_reader(reader, synthetic_dataset.data, 200)
+        list(reader)
```
this no longer checks the ids being read, I understand that this test is more to check that we can read with the disk cache without failures but is there a reason not to check that the ids make sense as well?

selitvin replied:
I removed it because _check_simple_reader does not support batches. Replaced it with the test on ids. That's concise enough.

selitvin replied:
Added an id check. Checking all fields is a bit trickier (though possible). Do you think it's needed?

rgruener replied:
I dont think its necessary. I think that this test doesnt need to cover that logic.

rgruener commented on petastorm/reader.py:
```diff
@@ -241,6 +346,18 @@ def __init__(self, pyarrow_filesystem, dataset_path, schema_fields=None,
+
+        :param worker_class: This is the class that will be instantiated on a different thread/process. It's
+            responsibility is to load and filter the data.
+        :param results_queue_reader: An instance of an object that will be reading data produced by worker class.
```
Since this is tightly coupled with the worker class does it make sense to have the results_queue_reader as a separate parameter?

selitvin replied:
Fixed.

selitvin commented:
Renamed `make_parquet_reader` -> `make_batch_reader`
Added explicit checks to these functions to forbid opening Petastorm stores from `make_batch_reader` and native Parquet stores using `make_reader`.

rgruener commented on petastorm/unischema.py:
```diff
@@ -206,7 +208,13 @@ def from_arrow_schema(cls, arrow_schema):
         unischema_fields = []
+
+        for partition_name in parquet_dataset.partitions.partition_names:
+            unischema_fields.append(UnischemaField(partition_name, np.str_, (), ScalarCodec(StringType()), False))
```
nice catch on the partition columns!
